pubsub/pubsub-pipe-image/pubsub-to-bigquery.py

```diff
@@ -0,0 +1,158 @@
+
+PUBSUB_TOPIC = os.environ['PUBSUB_TOPIC']
+
+PUBSUB_SCOPES = ['https://www.googleapis.com/auth/pubsub']
+NUM_RETRIES = 3
+
+
+def create_pubsub_client():
+    """Build the pubsub client."""
+    credentials = GoogleCredentials.get_application_default()
+    if credentials.create_scoped_required():
+            credentials = credentials.create_scoped(PUBSUB_SCOPES)
+    http = httplib2.Http()
+    credentials.authorize(http)
+    return discovery.build('pubsub', 'v1beta2', http=http)
+
+
+def fqrn(resource_type, project, resource):
+    """Returns a fully qualified resource name for Cloud Pub/Sub."""
+    return "projects/{}/{}/{}".format(project, resource_type, resource)
+
+
+def create_subscription(client, project_name, sub_name):
+    """Creates a new subscription to a given topic."""
+    print "using pubsub topic: %s" % PUBSUB_TOPIC
+    name = get_full_subscription_name(project_name, sub_name)
+    body = {'topic': PUBSUB_TOPIC}
+    subscription = client.projects().subscriptions().create(
+            name=name, body=body).execute(num_retries=NUM_RETRIES)
+    print 'Subscription {} was created.'.format(subscription['name'])
+
+
+def get_full_subscription_name(project, subscription):
+    """Returns a fully qualified subscription name."""
+    return fqrn('subscriptions', project, subscription)
+
+
+def pull_messages(client, project_name, sub_name):
+    """Pulls messages from a given subscription."""
+    BATCH_SIZE = 50
+    tweets = []
+    subscription = get_full_subscription_name(project_name, sub_name)
+    body = {
+            'returnImmediately': False,
+            'maxMessages': BATCH_SIZE
+    }
+    try:
+        resp = client.projects().subscriptions().pull(
+                subscription=subscription, body=body).execute(
+                        num_retries=NUM_RETRIES)
+    except Exception as e:
+        print "Exception: %s" % e
+        time.sleep(0.5)
+        return
+    receivedMessages = resp.get('receivedMessages')
+    if receivedMessages is not None:
+        ack_ids = []
+        for receivedMessage in receivedMessages:
+                message = receivedMessage.get('message')
+                if message:
+                        tweets.append(
+                            base64.b64decode(str(message.get('data'))))
+                        ack_ids.append(receivedMessage.get('ackId'))
+        ack_body = {'ackIds': ack_ids}
+        client.projects().subscriptions().acknowledge(
+                subscription=subscription, body=ack_body).execute(
+                        num_retries=NUM_RETRIES)
+    return tweets
+
+
+def create_bigquery_client():
+        """Build the bigquery client."""
+        credentials = GoogleCredentials.get_application_default()
+        if credentials.create_scoped_required():
+                credentials = credentials.create_scoped(BQ_SCOPES)
+        http = httplib2.Http()
+        credentials.authorize(http)
+        return discovery.build('bigquery', 'v2', http=http)
```

Comment:
Yeah, discovery.build will retrieve the discovery doc via the network.
If this is called often, you should cache the discovery doc, but if not, it might be OK.